cmd/write_graph.go

```diff
@@ -0,0 +1,205 @@
+
+func init() {
+	var parallelism int = 1
+	var startDelay int64 = 5
+	var number int64 = 1000000
+	cmdWriteGraph.Flags().IntVar(&parallelism, "parallelism", parallelism, "set -parallelism to use multiple go routines")
+	cmdWriteGraph.Flags().Int64Var(&number, "number", number, "set -number for number of edges to write per go routine")
+	cmdWriteGraph.Flags().Int64Var(&startDelay, "start-delay", startDelay, "Delay between the start of two go routines.")
+}
+
+// writeGraph writes edges in parallel
+func writeGraph(cmd *cobra.Command, _ []string) error {
+	parallelism, _ := cmd.Flags().GetInt("parallelism")
+	number, _ := cmd.Flags().GetInt64("number")
+	startDelay, _ := cmd.Flags().GetInt64("start-delay")
+
+	db, err := _client.Database(context.Background(), "_system")
+	if err != nil {
+		return errors.Wrapf(err, "can not get database: %s", "_system")
+	}
+
+	if err := writeGraphParallel(parallelism, number, startDelay, db); err != nil {
+		return errors.Wrapf(err, "can not setup some tenants")
+	}
+
+	return nil
+}
+
+// writeSomeGraphParallel creates some edges in parallel
+func writeGraphParallel(parallelism int, number int64, startDelay int64, db driver.Database) error {
+	var mutex sync.Mutex
+	totaltimestart := time.Now()
+	wg := sync.WaitGroup{}
+	haveError := false
+	for i := 1; i <= parallelism; i++ {
+	  time.Sleep(time.Duration(startDelay) * time.Millisecond)
+		i := i // bring into scope
+		wg.Add(1)
+
+		go func(wg *sync.WaitGroup, i int) {
+			defer wg.Done()
+			fmt.Printf("Starting go routine...\n")
+			id := "id_" + strconv.FormatInt(int64(i), 10)
+			err := writeSomeGraph(number, id, db, &mutex)
+			if err != nil {
+				fmt.Printf("writeSomeGraph error: %v\n", err)
+				haveError = true
+			}
+			mutex.Lock()
+			fmt.Printf("Go routine %d done\n", i)
+			mutex.Unlock()
+		}(&wg, i)
+	}
+
+	wg.Wait()
+	totaltimeend := time.Now()
+	totaltime := totaltimeend.Sub(totaltimestart)
+	docspersec := float64(int64(parallelism) * number) / (float64(totaltime) / float64(time.Second))
+	fmt.Printf("\nTotal number of edges written: %d, total time: %v, total edges per second: %f\n", int64(parallelism) * number, totaltimeend.Sub(totaltimestart), docspersec)
+	if !haveError {
+		return nil
+	}
+	fmt.Printf("Error in writeSomeGraph.\n")
+	return fmt.Errorf("Error in writeSomeGraph.")
+}
+
+// writeOneTenant does `nr` write operations, alternating between vertices
+// and edges and inserts and updates.
+func writeSomeGraph(nr int64, id string, db driver.Database, mutex *sync.Mutex) error {
+	instances, err := db.Collection(nil, "instances")
+	if err != nil {
+		fmt.Printf("writeSomeGraph: could not open `instances` collection: %v\n", err)
+		return err
+	}
+	steps, err := db.Collection(nil, "steps")
+	if err != nil {
+		fmt.Printf("writeSomeGraph: could not open `steps` collection: %v\n", err)
+		return err
+	}
+	optype := 0   // changes from 0 to 3 and then back to 0
+  times := make([]time.Duration, 0, 10000)
```

Comment:
here I can see that you are not using GO fmt tool, you can use one of these:
- go fmt <filename>
- in golang IDE setting -> watchers -> add go fmt (every time you save file it will format automatically for you)

Reply:
Done.